### PR TITLE
Refactoring of encode and decode to be zero-copy

### PR DESCRIPTION
**Pull request by mattiasdrp**

### CODE REVIEW

**Comment by samoht** [comment]
We should adapt this RFC to `repr`: https://github.com/ocaml/RFCs/blob/master/rfcs/modular_io.md

**Comment by mattiasdrp** [comment]
Encode and Decode have been implemented (decode not completely). Right now we need to make sure that these are the types we want and the direction we want to take.

A channel is potentially a stream without a fixed length so the unboxed version may be trickier to implement.

**Comment by samoht** [comment]
Before going further I think we should see what kind of impact this API will have on its users, eg. index and irmin. 

For instance, for both of these use-case the data to read and write will come through `pread` and `pwrite`, which require a byte buffer to be provided. So I think the main question to answer is where that buffer is allocated. I don't think your functor proposal solves this problem (but I might be missing something).

Also, is `decode` a bottleneck right now? https://github.com/mirage/index/issues/280 seems to only talk about `encode` so can we focus on this one first?  The first step is to identify how we can modify index to stop allocating intermediate buffers and directly write from the file buffer: if we'd want to do this, what API should `repr` provide? (e.g. write the user code first, before writing the actual implementation)

**Comment by pascutto** [comment]
> Also, is `decode` a bottleneck right now? [mirage/index#280](https://github.com/mirage/index/issues/280) seems to only talk about `encode` so can we focus on this one first?

Actually that's my mistake, I believe decoding is just as (if not more) important as encoding during merges, as every `data` entry needs to be decoded, but not all of them need to be encoded. I'll mention that back in https://github.com/mirage/index/issues/280.

**Comment by CraigFe** [comment]
To relate a discussion had with @mattiasdrp offline: the approach currently taken here isn't feasible as it forks the universe of typereps according to the channel types. (Repr.t values built for different functor instantiations would be incompatible, when the overall goal is to have a "canonical" type for typereps that can then be consumed uniformly.) This was motivated by wanting to make `Custom` extensible, but fortunately we don't need to solve this problem for Index / Irmin / Tezos as they use simple types -- we can just `raise Failure` in these cases for now, as `type_random.ml` does. (This extensibility issue is something that could be considered as part of a separate change.)

One path forward is to get a minimal working diff for an externally-allocated generic codec in Index (that simply fails in the corner cases), in order to demonstrate viable user code in Index and Irmin. As @samoht suggests, this is the main source of uncertainty (and fortunately these use-cases have simplifying assumptions that we can exploit for a proof of concept).

File: bench/main.ml [hunk 1]
```diff
@@ -1,6 +1,28 @@
 open Bechamel
 open Toolkit
 module T = Repr
+
+module IO = struct
```
[the remark this answers is not in the document]
Comment (mattiasdrp) [reply bench/main.ml:5]: Ok, I'm fine with that! Will do immediately :-)

**Comment by icristescu** [comment]
As discussed offline with @mattiasdrp we should focus on changing `encode` for now and see how that propagates to irmin/index. 

The API changes to `decode` consists of changing `string` with `bytes` (which is not changing much), but there are extra allocations done by `decode` (for instance https://github.com/mirage/repr/blob/main/src/repr/type_binary.ml#L261) that we can try to remove in a second step (a different PR?) as they are independent of the `encode` modifications.

File: bench/dune [hunk 2]
```diff
@@ -2,7 +2,7 @@
  (name main)
  (public_name repr-bench)
  (package repr-bench)
- (libraries repr bechamel fpath yojson unix)
+ (libraries repr bechamel fpath yojson unix memtrace)
```
Comment (icristescu) [inline bench/dune:5]: The ci fails because of this dependency, try adding it to the opam file too

File: src/repr/type.ml [hunk 3]
```diff
@@ -60,6 +39,31 @@ let triple a b c = Tuple (Triple (a, b, c))
...
+    | Self s -> aux s.self_fix
+    | Map m ->
+        let dst = unstage (aux m.x) in
+        stage (fun v -> dst (m.g v))
+    | Custom c -> c.pre_hash
+    | t -> Type_binary.Unboxed.encode_bin t
+  in
+  aux t
+
+let short_hash = function
```
Comment (icristescu) [inline src/repr/type.ml:53]: I don't think you need to move these functions (`pre_hash` and `short_hash`); its harder to review the difference like this.

File: src/repr/type_binary.ml [hunk 4]
```diff
@@ -19,112 +19,110 @@ open Staging
...
-    unsafe_add_bytes b
+  let char c byt off =
+    Bytes.set byt off c;
+    off + 1
 
-  let int32 i =
-    let b = Bytes.create 4 in
-    Bytes.set_int32_be b 0 i;
-    unsafe_add_bytes b
+  let byte n byt off = char (Char.chr (0xFF land n)) byt off
```
Comment (icristescu) [inline src/repr/type_binary.ml:38]: I think I missed something. why are you doing `0xFF land n`

File: src/repr/type_binary.ml [hunk 5]
```diff
@@ -19,112 +19,110 @@ open Staging
...
+  (* let unsafe_add_bytes b oc = IO.append_bytes oc b *)
```
Comment (icristescu) [inline src/repr/type_binary.ml:39]: comment forgot here

File: src/repr/type_binary.ml [hunk 6]
```diff
@@ -19,112 +19,110 @@ open Staging
...
-  let int i k =
-    let rec aux n k =
-      if n >= 0 && n < 128 then k chars.(n)
-      else
-        let out = 128 lor (n land 127) in
-        k chars.(out);
-        aux (n lsr 7) k
-    in
-    aux i k
+  let rec aux n byt off =
```
Comment (icristescu) [inline src/repr/type_binary.ml:60]: maybe rename `aux` into something more meaningful now that it is outside of `int`?

File: src/repr/type_binary.ml [hunk 7]
```diff
@@ -19,112 +19,110 @@ open Staging
...
+    if n >= 0 && n < 128 then byte n byt off
+    else
+      let out = 128 lor (n land 127) in
+      byte out byt off |> aux (n lsr 7) byt
 
-  let len n i =
+  let int = aux
+
+  let len n i byt off =
```
Comment (icristescu) [inline src/repr/type_binary.ml:68]: I don't think you need to modify this function

File: src/repr/type_binary.ml [hunk 8]
```diff
@@ -19,112 +19,110 @@ open Staging
...
-        | Some x ->
-            char '\255' k;
-            o x k)
+        | None -> char '\000' byt off
+        | Some x -> char '\255' byt off |> o x byt)
 
-  let rec t : type a. a t -> a encode_bin = function
+  let rec t : type a. a t -> a encode_bin =
+   fun ty ->
+    match ty with
```
Comment (icristescu) [inline src/repr/type_binary.ml:125]: you don't need to modify these lines

File: src/repr/type_binary.ml [hunk 9]
```diff
@@ -165,7 +163,7 @@ module Encode = struct
   and map : type a b. boxed:bool -> (a, b) map -> b encode_bin =
    fun ~boxed { x; g; _ } ->
     let encode_bin = unstage (if boxed then t x else unboxed x) in
-    stage (fun u k -> encode_bin (g u) k)
+    stage (fun u oc -> encode_bin (g u) oc)
```
Comment (icristescu) [inline src/repr/type_binary.ml:166]: `encode_bin` now expects 3 arguments - for consistency can you call them `byt` and `off` as below?

File: src/repr/type_binary.ml [hunk 10]
```diff
@@ -425,15 +422,14 @@ module Unboxed = struct
   let decode_bin = Decode.unboxed
 end
 
-let to_bin size_of encode_bin =
+let to_bin size_of (encode_bin : 'a encode_bin) =
```
Comment (icristescu) [inline src/repr/type_binary.ml:425]: you don't need this type annotation I think

File: test/repr/main.ml [hunk 11]
```diff
@@ -20,11 +20,15 @@ let of_bin_string t = T.unstage (T.of_bin_string t)
...
 let decode_bin t = T.unstage (T.decode_bin t)
 let size_of t = T.unstage (T.size_of t)
+let sub_bytes len off byt = if len = off then byt else Bytes.sub byt 0 off
+let sub_string len off byt = Bytes.to_string @@ sub_bytes len off byt
 
-let with_buf f =
-  let buf = Buffer.create 10 in
-  f (Buffer.add_string buf);
-  Buffer.contents buf
+let with_byt ty v size_of f =
```
Comment (icristescu) [inline test/repr/main.ml:26]: nitpicking, but do you mind renaming the function into `with_bytes` and `ty` to `t` for consistency

File: test/repr/main.ml [hunk 12]
```diff
@@ -240,22 +244,28 @@ let l =
...
+let to_bytes ty v f size_of =
+  let len = match (size_of ty) v with None -> 1024 | Some n -> n in
+  let byt = Bytes.create len in
+  let off = f ty v byt 0 in
+  sub_bytes len off byt
+
 let test_bin () =
-  let s = T.to_string l [ "foo"; "foo" ] in
-  Alcotest.(check string) "hex list" "[\"666f6f\",\"666f6f\"]" s;
+  let s = T.to_string l [ "foo"; "bar" ] in
```
Comment (icristescu) [inline test/repr/main.ml:254]: maybe don't change the test?

File: src/repr/type_binary.ml (hunk 4 above)
Comment (mattiasdrp) [reply src/repr/type_binary.ml]: A lot of the things you noticed were things I forgot to clean, thanks and sorry for making you do this :)

**Pull request closed without merging**